### PR TITLE
Migrate to micrometer metrics

### PR DESCRIPTION
**Pull request by bennetelli**
the build takes hours.. I am expecting that there are still some failing tests.

### CODE REVIEW

File: external/src/test/java/com/redhat/cloud/policies/engine/process/ReceiverTest.java [hunk 1]
```diff
@@ -127,7 +128,7 @@ private Action deserializeAction(String payload) {
             DatumReader<Action> reader = new SpecificDatumReader<>(Action.class);
             reader.read(action, jsonDecoder);
         } catch (IOException e) {
-            throw new IllegalArgumentException("Payload extraction failed", e);
+            throw new IllegalArgumentException("Payload extractioreturn null;n failed", e);
```
Comment (theute) [inline external/src/test/java/com/redhat/cloud/policies/engine/process/ReceiverTest.java:131]: Doesn't sound right ;-)

Comment (bennetelli) [reply external/src/test/java/com/redhat/cloud/policies/engine/process/ReceiverTest.java:131]: Oops. Copy paste issue.

**Comment by bennetelli** [comment]
in case evey test is green, it could still be that the micrometer metrics look a bit different than the microprofile ones. Probably need to change grafana queries as well.

**Comment by bennetelli** [comment]
> Does that replace #255 ? (If so, I guess 255 should be closed ?)

No, this task is the same as in #255, but for policies-engine. There is a jira ticket missing for it. 
Was diving into the code because of the prod issue and wanted to improve things while getting into it.

**Comment by bennetelli** [comment]
The tests are green. I can't believe it..

**Comment by bennetelli** [comment]
Maybe I am a bit in a tunnel. don't find something to change. Maybe you find something? :)

File: external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java [hunk 2]
```diff
@@ -315,14 +315,15 @@ public void createFullTrigger(RoutingContext routing) {
...
+    @Timed
+            //(absolute = true, tags = {"path=/triggers/trigger/{triggerId}", "method=PUT"})
     public void updateFullTrigger(RoutingContext routing) {
         routing.vertx()
                 .executeBlocking(future -> {
                     String tenantId = ResponseUtil.checkTenant(routing);
 
                     // TODO Temp counter used to investigate the engine instability, remove ASAP.
-                    metricRegistry.counter("policies_engine_updateFullTrigger", new Tag("account", tenantId)).inc();
+                    meterRegistry.counter("policies_engine_updateFullTrigger", "account", tenantId).increment();
```
Comment (gwenneg) [inline external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java:326]: General comment: will the dynamic tags be available in Prometheus? It didn't work in notifications.

File: external/src/main/java/com/redhat/cloud/policies/engine/process/Receiver.java [hunk 3]
```diff
@@ -82,46 +84,40 @@ public class Receiver {
...
+    Counter processingErrors;
 
     @Inject
-    @Metric(absolute = true, name = "engine.input.processed.errors", tags = {"queue=host-egress"})
-    Counter processingErrors;
+    MeterRegistry meterRegistry;
+
+    @PostConstruct
+    public void init() {
+        incomingMessagesCount = meterRegistry.counter("engine.input.processed");
```
Comment (gwenneg) [inline external/src/main/java/com/redhat/cloud/policies/engine/process/Receiver.java:100]:
```suggestion
        incomingMessagesCount = meterRegistry.counter("engine.input.processed", "queue", "host-egress");
```

File: external/src/main/java/com/redhat/cloud/policies/engine/process/Receiver.java [hunk 4]
```diff
@@ -82,46 +84,40 @@ public class Receiver {
...
+        rejectedCount = meterRegistry.counter("engine.input.rejected");
```
Comment (gwenneg) [inline external/src/main/java/com/redhat/cloud/policies/engine/process/Receiver.java:101]:
```suggestion
        rejectedCount = meterRegistry.counter("engine.input.rejected", "queue", "host-egress");
```

File: external/src/main/java/com/redhat/cloud/policies/engine/metrics/ProcSelfStatusExporter.java [hunk 5]
```diff
@@ -33,15 +36,31 @@ public class ProcSelfStatusExporter {
...
+    public void initCounters() {
+        meterRegistry.gauge("status.vmHwm", vmHwm);
+        meterRegistry.gauge("status.vmRss", vmRss);
+        meterRegistry.gauge("status.rssAnon", rssAnon);
+        meterRegistry.gauge("status.rssFile", rssFile);
+        meterRegistry.gauge("status.vmStk", vmStk);
+        meterRegistry.gauge("status.vmLib", vmLib);
+        meterRegistry.gauge("status.vmData", vmData);
+        meterRegistry.gauge("status.vmSize", vmSize);
+        meterRegistry.gauge("status.threads", threads);
```
Comment (gwenneg) [inline external/src/main/java/com/redhat/cloud/policies/engine/metrics/ProcSelfStatusExporter.java:62]: The old tags are missing. Units too but I don't know if Micrometer supports that.

File: external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java [hunk 6]
```diff
@@ -220,14 +219,15 @@ void createDampening(RoutingContext routing, boolean isGroup) {
             @DocResponse(code = 400, message = "Bad Request/Invalid Parameters.", response = ResponseUtil.ApiError.class),
             @DocResponse(code = 500, message = "Internal server error.", response = ResponseUtil.ApiError.class)
     })
-    @Timed(absolute = true, tags = {"path=/triggers/trigger", "method=POST"})
+    @Timed
+            //(tags = {"path=/triggers/trigger", "method=POST"})
```
Comment (gwenneg) [inline external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java:223]: Are the old tags automatically added by Micrometer? Should that comment be removed?

File: external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java [hunk 7]
```diff
@@ -315,14 +315,15 @@ public void createFullTrigger(RoutingContext routing) {
             @DocResponse(code = 400, message = "Bad Request/Invalid Parameters.", response = ResponseUtil.ApiError.class),
             @DocResponse(code = 500, message = "Internal server error.", response = ResponseUtil.ApiError.class)
     })
-    @Timed(absolute = true, tags = {"path=/triggers/trigger/{triggerId}", "method=PUT"})
+    @Timed
+            //(absolute = true, tags = {"path=/triggers/trigger/{triggerId}", "method=PUT"})
```
Comment (gwenneg) [inline external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java:319]: Are the old tags automatically added by Micrometer? Should that comment be removed?

File: external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java [hunk 8]
```diff
@@ -657,14 +658,15 @@ public void deleteGroupTrigger(RoutingContext routing) {
             @DocResponse(code = 404, message = "Trigger not found", response = ResponseUtil.ApiError.class),
             @DocResponse(code = 500, message = "Internal server error.", response = ResponseUtil.ApiError.class)
     })
-    @Timed(absolute = true, tags = {"path=/triggers/{triggerId}", "method=DELETE"})
+    @Timed
+            //(absolute = true, tags = {"path=/triggers/{triggerId}", "method=DELETE"})
```
Comment (gwenneg) [inline external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java:662]: Are the old tags automatically added by Micrometer? Should that comment be removed?

File: external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java [hunk 9]
```diff
@@ -739,15 +741,16 @@ public void findGroupMembers(RoutingContext routing) {
             @DocResponse(code = 404, message = "Trigger not found", response = ResponseUtil.ApiError.class),
             @DocResponse(code = 500, message = "Internal server error.", response = ResponseUtil.ApiError.class)
     })
-    @Timed(absolute = true, tags = {"path=/triggers", "method=GET"})
+    @Timed
+            //(absolute = true, tags = {"path=/triggers", "method=GET"})
```
Comment (gwenneg) [inline external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java:745]: Are the old tags automatically added by Micrometer? Should that comment be removed?

File: external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java [hunk 10]
```diff
@@ -839,7 +842,8 @@ public void getTrigger(RoutingContext routing) {
             @DocResponse(code = 404, message = "Trigger not found", response = ResponseUtil.ApiError.class),
             @DocResponse(code = 500, message = "Internal server error.", response = ResponseUtil.ApiError.class)
     })
-    @Timed(absolute = true, tags = {"path=/triggers/trigger/{triggerId}", "method=GET"})
+    @Timed
+            //(absolute = true, tags = {"path=/triggers/trigger/{triggerId}", "method=GET"})
```
Comment (gwenneg) [inline external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java:846]: Are the old tags automatically added by Micrometer? Should that comment be removed?

File: external/src/test/java/com/redhat/cloud/policies/engine/process/ReceiverFilterTest.java [hunk 11]
```diff
@@ -118,31 +129,55 @@ public void testRejectionTypes() throws Exception {
...
+
+        @Override
+        public void use(Consumer<Gauge> visitGauge, Consumer<Counter> visitCounter, Consumer<Timer> visitTimer, Consumer<DistributionSummary> visitSummary, Consumer<LongTaskTimer> visitLongTaskTimer, Consumer<TimeGauge> visitTimeGauge, Consumer<FunctionCounter> visitFunctionCounter, Consumer<FunctionTimer> visitFunctionTimer, Consumer<Meter> visitMeter) {
+            Counter.super.use(visitGauge, visitCounter, visitTimer, visitSummary, visitLongTaskTimer, visitTimeGauge, visitFunctionCounter, visitFunctionTimer, visitMeter);
+        }
+
+        @Override
+        public void close() {
+            Counter.super.close();
+        }
```
Comment (gwenneg) [inline external/src/test/java/com/redhat/cloud/policies/engine/process/ReceiverFilterTest.java:176]: Is this code really needed? These 4 methods seem to already exist as `default` interface methods in `Meter` and `Counter`.

File: external/src/test/java/com/redhat/cloud/policies/engine/process/ReceiverTest.java [hunk 12]
```diff
@@ -163,6 +160,7 @@ public void resetWebhookReceiver() {
 
     @Test
     public void testReceiver() throws Exception {
+        meterRegistry.counter("engine.input.processed.errors", "queue", "host-egress");
```
[the remark this answers is not in the document]
Comment (bennetelli) [reply external/src/test/java/com/redhat/cloud/policies/engine/process/ReceiverTest.java:163]: Before it was an attribute and declared in the init method like this: errorCount = meterRegistry.counter("engine.input.processed.errors", "queue", "host-egress");

but it was used in one single method only. So I moved it in that method.

File: external/src/test/java/com/redhat/cloud/policies/engine/process/ReceiverTest.java [hunk 13]
```diff
@@ -107,18 +106,16 @@ void init() throws InterruptedException {
...
-    }
-
     private static final String TENANT_ID = "integration-test";
     private static final String ACTION_PLUGIN = "email";
     private static final String ACTION_ID = "email-notif";
     private static final String TRIGGER_ID = "arch-trigger";
 
-    private final MetricID errorCount = new MetricID("engine.input.processed.errors", new org.eclipse.microprofile.metrics.Tag("queue", "host-egress"));
+    @PostConstruct
+    public void initReactiveMessagingChannelsAndCounter() {
```
[the remark this answers is not in the document]
Comment (bennetelli) [reply external/src/test/java/com/redhat/cloud/policies/engine/process/ReceiverTest.java:115]: yea. the init method doesn't contain channels only anymore. That's why I thought about naming it just "init" or the name I am using right now.

File: external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java [hunk 14]
```diff
@@ -220,14 +219,15 @@ void createDampening(RoutingContext routing, boolean isGroup) {
...
+    @Timed
+            //(tags = {"path=/triggers/trigger", "method=POST"})
     public void createFullTrigger(RoutingContext routing) {
         routing.vertx()
                 .executeBlocking(future -> {
                     String tenantId = ResponseUtil.checkTenant(routing);
 
                     // TODO Temp counter used to investigate the engine instability, remove ASAP.
-                    metricRegistry.counter("policies_engine_createFullTrigger", new Tag("account", tenantId)).inc();
+                    meterRegistry.counter("policies_engine_createFullTrigger", "account", tenantId).increment();
```
[the remark this answers is not in the document]
Comment (bennetelli) [reply external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java]: replaced underscores with dots and counter with timer

File: external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java (hunk 2 above)
Comment (bennetelli) [reply external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java]: using timer now

File: external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java (hunk 7 above)
Comment (bennetelli) [reply external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java]: I tried to somehow add the tags with the last cmmit, but absolute = true isn't available

File: external/src/test/java/com/redhat/cloud/policies/engine/process/ReceiverFilterTest.java (hunk 11 above)
Comment (bennetelli) [reply external/src/test/java/com/redhat/cloud/policies/engine/process/ReceiverFilterTest.java]: removed

File: external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java (hunk 8 above)
Comment (bennetelli) [reply external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java]: I tried to somehow add the tags with the last cmmit, but absolute = true isn't available

File: external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java (hunk 9 above)
Comment (bennetelli) [reply external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java]: I tried to somehow add the tags with the last commit, but absolute = true isn't available

File: external/pom.xml [hunk 15]
```diff
@@ -82,8 +78,8 @@
 
     <!-- Quarkus related -->
     <dependency>
-      <groupId>org.eclipse.microprofile.metrics</groupId>
-      <artifactId>microprofile-metrics-api</artifactId>
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-micrometer-registry-prometheus</artifactId>
```
[the remark this answers is not in the document]
Comment (bennetelli) [reply external/pom.xml:80]: moved it one "section" up

File: external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java (hunk 10 above)
Comment (bennetelli) [reply external/src/main/java/com/redhat/cloud/policies/engine/handlers/TriggersHandler.java]: I tried to somehow add the tags with the last commit, but absolute = true isn't available

File: external/src/test/java/com/redhat/cloud/policies/engine/process/ReceiverTest.java [hunk 16]
```diff
@@ -207,7 +205,7 @@ public void testReceiver() throws Exception {
         checkPoliciesHistoryEntries(fullTrigger, 2);
         checkPoliciesHistoryEntries(fullTrigger2, 2);
 
-        Counter hostEgressProcessingErrors = meterRegistry.find(errorCount.getName()).counter();
+        Counter hostEgressProcessingErrors = meterRegistry.find("engine.input.processed.errors").counter();
```
[the remark this answers is not in the document]
Comment (bennetelli) [reply external/src/test/java/com/redhat/cloud/policies/engine/process/ReceiverTest.java:208]: how do you mean? I can only find metrics by their name as String.

**Comment by gwenneg** [comment]
Squashed, rebased on `upstream/master` and fixed conflicts.

**Review by gwenneg — APPROVED** [review]
LGTM

**Pull request merged**